lib/plugins/paragon-webpack-plugin/ParagonWebpackPlugin.js

```diff
@@ -0,0 +1,141 @@
+const { Compilation, sources } = require('webpack');
+const fs = require('fs');
+const path = require('path');
+const {
+  getParagonVersion,
+  getParagonThemeCss,
+} = require('../../../config/data/paragonUtils');
+const {
+  injectMetadataIntoDocument,
+  getParagonStylesheetUrls,
+  injectParagonCoreStylesheets,
+  injectParagonThemeVariantStylesheets,
+} = require('./utils');
+
+// Resolve configuration `env.config`.
+const envConfigFile = path.resolve(process.cwd(), 'env.config.js');
```

Comment:
[nit/inform] likely OK for now, but we might want to follow-up here to ensure we support `env.config` files defined with the `.jsx` extension (related to https://github.com/openedx/frontend-build/issues/514).

Reply:
I agree with this and I tried to use `.jsx` files,  and found the following scenario:

Due to Node only supporting js, json natively,  read jsx should be done with fs.readFile and processing the result (string) to find the `PARAGON_THEME_URLS`. 

The reason why I don't use it is that I don't want to increase the build time more processing the file.

Reply:
There are MFEs that currently assume the `.jsx` extension works and uses it ([source](https://github.com/search?q=org%3Aopenedx%20env.config.jsx&type=code)), as well as private `env.config.jsx` files currently used to introduce private JSX-based plugins/widgets (e.g., for at least 2U's instance of `frontend-app-learner-record`).

Given this, I'd be concerned that without `ParagonWebpackPlugin` supporting `.jsx`, any MFEs that define custom JSX-based plugins (e.g., via `@openedx/frontend-plugin-framework`) with `env.config.jsx` won't be able to configure `PARAGON_THEME_URLS`.

FWIW, an alternative to parsing the `env.config.jsx` as a string with `fs.readFile` would be to conditionally rely on `@babel/register` to use Babel to transform JSX to vanilla JS during the `require` import.

**Example solution with `@babel/register`**

1. Install `@babel/register` in root `package.json`.

2. Define a file `register-babel.js` with the following contents (e.g., under `lib` directory):

```js
require('@babel/register')({
  presets: ['@babel/preset-env', '@babel/preset-react'],
});
```

3. Update `ParagonWebpackPlugin` to check against supported extensions (`js` and `jsx`).

```js
const supportedEnvConfigExtensions = ['js', 'jsx'];
const envConfigFile = path.resolve(process.cwd(), 'env.config');
```

```js
async resolveParagonThemeUrlsFromConfig() {
  if (!supportedEnvConfigExtensions.some(ext => fs.existsSync(`${envConfigFile}.${ext}`))) {
    // env.config file does not exist
    return;
  }

  let envConfig;
  require('../../register-babel');
  try {
    envConfig = require(envConfigFile);
  } catch (error) {
    console.info('ParagonWebpackPlugin cannot load env.config file. Skipping...');
  }

  // ...
}
```

Reply:
> The reason why I don't use it is that I don't want to increase the build time more processing the file.

If existing MFEs / frontend-platform is compatible with JSX-based `env.config` files, I feel all usages of `env.config` should support the same extensions, even if it means slightly more processing at build time.

Otherwise, it's fairly confusing for consumers/engineers to know which `env.config` file extensions work for different parts of the system (e.g., can use JSX-based `env.config` for private plugins but not for `PARAGON_THEME_URLS`).

Reply:
Just to be clear `PARAGON_THEME_URLS` can still be implemented in  JSX-based configuration because `frontend-platform`  can read it so consumers/engineers can use it.

Thanks for the explanation about `register-babel` I will take a look.

Reply:
> Just to be clear PARAGON_THEME_URLS can still be implemented in JSX-based configuration because frontend-platform can read it so consumers/engineers can use it.

While true, not supporting JSX-based `env.config` in frontend-build comes with a performance tradeoff for users. If frontend-build does not support JSX-based `env.config` for defining `PARAGON_THEME_URLS`, the associated `<link rel="preload">` elements added to the HTML document by frontend-build would be missing from the HTML document.

Without the preloaded stylesheets ([reference about preload](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/preload)), users will need to wait to download/parse/execute the MFE's JS (including `@edx/frontend-platform`'s source) _before_ the `<link>` elements would be created / download started for the configured `PARAGON_THEME_URLS` defined by `env.config.jsx`.

Given `AppProvider` in frontend-platform blocks the rendering of the MFE's contents until the Paragon CSS is loaded ([source](https://github.com/openedx/frontend-platform/pull/689/files#diff-e986d374444ced0f5162cc6cb8b3dac6cbd228f339851f2a99a4aec545559d87R89-R91)), I feel it's important that these `<link>` elements are preloaded on initial page load regardless of using `env.config.js` or `env.config.jsx`, without being blocked by downloading JavaScript assets first to ensure this solution is as performant as possible for users.

Reply:
I have been testing `@babel/register` and works when the file is `jsx` but not alongside `@openedx/frontend-plugin-framework` (fail with `SyntaxError: Unexpected token 'export'` trying to resolve the plugin dependencies). 

I am testing if some plugin can help in that case

Reply:
I couldn't find a solution for the dependencies issue, @adamstankiewicz do you have an idea that could work? I'm using the file in learning MFE for testing leaving only `@openedx/frontend-plugin-framework` the example for footer-slot and PARAGON_THEME_URLS

```jsx
import { PLUGIN_OPERATIONS, DIRECT_PLUGIN } from '@openedx/frontend-plugin-framework';

// Load environment variables from .env file
const config = {
  ...process.env,
  pluginSlots: {
    footer_slot: {
      plugins: [
        {
          // Hide the default footer
          op: PLUGIN_OPERATIONS.Hide,
          widgetId: 'default_contents',
        },
        {
          // Insert a custom footer
          op: PLUGIN_OPERATIONS.Insert,
          widget: {
            id: 'custom_footer',
            type: DIRECT_PLUGIN,
            RenderWidget: () => (
              <h1 style={{ textAlign: 'center' }}>🦶</h1>
            ),
          },
        },
      ]
    }
  },
  PARAGON_THEME_URLS: {
    core: {
      urls: {
        default: 'https://cdn.jsdelivr.net/npm/@openedx/paragon@alpha/dist/core.min.css',
      },
    },
    defaults: {
      light: 'light',
    },
    variants: {
      light: {
        urls: {
          default: 'https://cdn.jsdelivr.net/npm/@openedx/paragon@alpha/dist/light.min.css',
        },
      },
    },
  },
};

export default config;
```

The import error persists.

Reply:
@dcoa do you have a branch or a diff somewhere with what you've tried in regards to `@babel/register`? I find it's often easier to talk about code when there's a diff to look at.

Also, I did a bit of searching and I wasn't able to find much about the error you're seeing outside of https://stackoverflow.com/a/54663822. Someone resolved a similar issue by setting a root directory for babel, but I'm not sure if that applies at all here.

Reply:
[Just for the record] [This one](https://gist.github.com/dcoa/f53caa609fd9f817c87baee038e4a528) compiles the different config that I use to test `@babel/register`

Reply:
@adamstankiewicz @brian-smith-tcril  as mentioned in Slack, I added the proposed workaround to support `process.env.PARAGON_THEME_URLS`. 

We can storage the variable in JSON format but should be a minified version of it (1 line)

```env
PARAGON_THEME_URLS= {core:{urls:{default:'https://cdn.jsdelivr.net/npm/@openedx/paragon@alpha/dist/core.min.css'}},defaults:{light:'light'},variants:{light:{urls:{default:'https://cdn.jsdelivr.net/npm/@openedx/paragon@alpha/dist/light.min.css'}}}}

```